File: leetcodePractice/6-25-2020/permutation.java
/*
全排列，lintcode-15
 */

public List<List<Integer>> permute(int[] nums) {
    List<List<Integer>> permutations = new ArrayList<>();
    if (nums == null) {
        return permutations;
    }
    if (nums.length == 0) {
        permutations.add(new ArrayList<Integer>());
        return permutations;
    }

    boolean[] visited = new boolean[nums.length];
    List<Integer> permutations = new ArrayList<>();
    dfs(nums, visited, permutation, permutations);

    return permutations;
}

// 递归的定义：找到以permutation开头的所有排列
private void dfs(int[] nums,
                 boolean[] visited,
                 List<Integer> permutation,
                 List<List<Integer>> permutations){
    // 递归的出口：当当前排列的长度等于集合长度
    if (permutation.size() == nums.length) {
        permutations.add(new ArrayList<Integer>(permutation));
        return;
    }

    // 递归的拆解
    // [] -> [1], [2], [3] ...
    // [1] -> [1,2], [1,3], [1,4] ...
    for (int i = 0; i < nums.length; i++) {
        if (visited[i]) {
            continue;
        }
        visited[i] = true;
        permutation.add(nums[i]);
        dfs(nums, visited, permutation, permutations);
        permutation.remove(permutation.size() - 1); //backtracking
        visited[i] = false; // backtracking
    }
}

File: leetcodePractice/6-25-2020/permutation.py
'''
全排列
'''

def permute(self, nums):
    if not nums:
        # 0!= 1
        return [[]]

    permutations = []
    self.dfs(nums, set(), [], permutations)
    return permutations

# 递归的定义： 找到所有 permutation 开头的permutations
def dfs(self, nums, visited, permutation, permutations):
    # 递归的出口
    if len(nums) == len(permutation):
        permutations.append(list(permutation))
        return
    
    # 递归的拆解
    # [] -> [1], [2], [3]
    # [1] -> [1,2], [1,3], [1,4]
    for num in nums:
        if num in visited:
            continue
        permutation.append(num)
        visited.add(num)
        self.dfs(nums, visited, permutation, permutations)
        visited.remove(num)
        permutation.pop()